### PR TITLE
Enforce both input tensor shapes of CosineEmbeddingLoss to be equal. 

### LINKED ISSUE #112732: Segmentation fault in torch.nn.functional.cosine_embedding_loss with empty tensors and mixed dtypes
### 🐛 Describe the bug

When calling `torch.nn.functional.cosine_embedding_loss` with empty tensors and a mix of float16, int8, and float32 data types, a segmentation fault occurs. Additionally, a deprecation warning for the size_average and reduce arguments is triggered, which does not halt execution but indicates that these arguments will be deprecated.
**To reproduce**
```python
import torch
import numpy as np

args = {
    'input1': torch.from_numpy(np.array(np.random.uniform(-1000_000, 1000_000, [0, 0])).astype('float16')), 
    'input2': torch.from_numpy(np.array(np.random.randint(-1000_000, 1000_000, [0])).astype('int8')),
    'margin': 6.4513009, 
    'reduce': True, 
    'reduction': 'mean', 
    'size_average': True, 
    'target': torch.from_numpy(np.array(np.random.uniform(-1000_000, 1000_000, [0])).astype('float32')), 
}

torch.nn.functional.cosine_embedding_loss(**args)
```

**Error messages and stack trace**
```
/home/guihuan/.conda/envs/py39/lib/python3.9/site-packages/torch/nn/_reduction.py:42: UserWarning: size_average and reduce args will be deprecated, please use reduction='mean' instead.
  warnings.warn(warning.format(ret))
Segmentation fault (core dumped)
```

### Versions

PyTorch version: 2.2.0.dev20230921+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.3 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.9.17 (main, Jul 5 2023, 20:41:20) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.15.0-86-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.1.105
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration:
GPU 0: NVIDIA GeForce RTX 2070
GPU 1: NVIDIA GeForce RTX 2070
GPU 2: NVIDIA GeForce RTX 2070
GPU 3: NVIDIA GeForce RTX 2070

Nvidia driver version: 535.104.12
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture: x86_64
CPU op-mode(s): 32-bit, 64-bit
Address sizes: 46 bits physical, 48 bits virtual
Byte Order: Little Endian
CPU(s): 32
On-line CPU(s) list: 0-31
Vendor ID: GenuineIntel
Model name: Intel(R) Xeon(R) CPU E5-2630 v3 @ 2.40GHz
CPU family: 6
Model: 63
Thread(s) per core: 2
Core(s) per socket: 8
Socket(s): 2
Stepping: 2
CPU max MHz: 3200.0000
CPU min MHz: 1200.0000
BogoMIPS: 4794.64
Flags: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm cpuid_fault epb invpcid_single pti ssbd ibrs ibpb stibp tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm xsaveopt cqm_llc cqm_occup_llc dtherm ida arat pln pts md_clear flush_l1d
Virtualization: VT-x
L1d cache: 512 KiB (16 instances)
L1i cache: 512 KiB (16 instances)
L2 cache: 4 MiB (16 instances)
L3 cache: 40 MiB (2 instances)
NUMA node(s): 2
NUMA node0 CPU(s): 0,2,4,6,8,10,12,14,16,18,20,22,24,26,28,30
NUMA node1 CPU(s): 1,3,5,7,9,11,13,15,17,19,21,23,25,27,29,31
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit: KVM: Mitigation: VMX disabled
Vulnerability L1tf: Mitigation; PTE Inversion; VMX conditional cache flushes, SMT vulnerable
Vulnerability Mds: Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Meltdown: Mitigation; PTI
Vulnerability Mmio stale data: Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Retbleed: Not affected
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1: Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2: Mitigation; Retpolines, IBPB conditional, IBRS_FW, STIBP conditional, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds: Not affected
Vulnerability Tsx async abort: Not affected

Versions of relevant libraries:
[pip3] numpy==1.24.3
[pip3] pytorch-triton==2.1.0+6e4932cda8
[pip3] torch==2.2.0.dev20230921+cu118
[pip3] torchaudio==2.2.0.dev20230921+cu118
[pip3] torchvision==0.17.0.dev20230921+cu118
[conda] cudatoolkit 11.8.0 h6a678d5_0 defaults
[conda] numpy 1.24.3 pypi_0 pypi
[conda] pytorch-triton 2.1.0+6e4932cda8 pypi_0 pypi
[conda] torch 2.2.0.dev20230921+cu118 pypi_0 pypi
[conda] torchaudio 2.2.0.dev20230921+cu118 pypi_0 pypi
[conda] torchvision 0.17.0.dev20230921+cu118 pypi_0 pypi

cc @ezyang @gchanan @zou3519 @kadeng @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki

### PR DESCRIPTION
**Pull request by tringwald**
…Added a test to prevent regressions.

Fixes #112732.

### CODE REVIEW

File: aten/src/ATen/native/Loss.cpp [hunk 1]
```diff
@@ -146,6 +146,13 @@ Tensor cosine_embedding_loss(const Tensor& input1, const Tensor& input2, const T
   TORCH_CHECK(
       targ_dim == 1 || targ_dim == 0,
       "0D or 1D target tensor expected, multi-target not supported");
+  TORCH_CHECK(
+      input1.sizes() == input2.sizes(),
+      "Expected input1 to be of same shape as input2, but got ",
+      input1.sizes(),
```
Comment (mikaylagawarecki) [inline aten/src/ATen/native/Loss.cpp:152]: You might need to change these to `sym_sizes()` to fix the failing CI

**Review by lezcano — COMMENTED** [review]
The previous code did allow broadcasting. Perhaps we want to allow broadcasting as well?

**Comment by tringwald** [comment]
> The previous code did allow broadcasting. Perhaps we want to allow broadcasting as well?

I've changed the checking logic a bit. Looks like we just need to make sure the second tensor has the same number of dimensions as the first one. Things like this never worked anyway:
```python3
>>> torch.__version__
'2.1.0'
>>> torch.nn.functional.cosine_embedding_loss(torch.ones(2, 10), torch.ones(10), target=torch.ones(1))
[...]
IndexError: Dimension out of range (expected to be in range of [-1, 0], but got 1)
>>> torch.nn.functional.cosine_embedding_loss(torch.ones(10), torch.ones(2, 10), target=torch.ones(1))
RuntimeError: 1D target tensor expects 2D input tensors, but found inputs with sizes [10] and [2, 10].
```

Should we mention broadcasting semantics in the docs too? Right now it says [Input2: (N,D) or (D), same shape as Input1.](https://pytorch.org/docs/stable/generated/torch.nn.CosineEmbeddingLoss.html)

**Review by lezcano — APPROVED** [review]
LGTM. Updating the docs to shortly mention broadcasting would also be good.

**Pull request closed without merging**